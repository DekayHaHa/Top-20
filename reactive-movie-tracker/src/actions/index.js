export const addMovies = (movies) => ({
	type: 'ADD_MOVIES',
	movies
})

export const signInUser = (id, name) => ({
	type: 'SIGNIN_USER',
	name,
	id
})

export const signOutUser = () => ({
	type: 'SIGN_OUT_USER',
})

export const isLoading = (bool) => ({
	type: 'IS_LOADING',
	isLoading: bool
})

export const hasError = (message) => ({
	type: 'HAS_ERROR',
	message
})