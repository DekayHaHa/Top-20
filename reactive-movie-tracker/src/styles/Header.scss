.Header {
  display: flex;
  background-color: gray;
  justify-content: space-between;
  align-items: center;
  color: salmon;
  border-style: solid;
  border-color: gray;
  border-radius: 5px; 
}

.btn {
  background: salmon;
  color: #fff;
  border-radius: 5px;
  font-size: 15px;
  padding: 1em;
  margin: 1em 20px;
  box-shadow: 0px 1px 2px 2px black;
  border: none;
  font-weight: 100;
  cursor: pointer;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;

  &:hover {
    background: #000000;
  }

}

.title {
  font-size: 25px;
  margin: 5px
}