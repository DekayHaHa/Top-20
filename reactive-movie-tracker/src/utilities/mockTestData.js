export const mockReduxMovies = [
	{	title: 'Kung Fu',
	image: 'www.reddit.com',
	overview: 'Big Blue House',
	score: 7.8,
	id: 166428,
	isFavorite: false,
	releaseDate: "2019-01-03",},
	{	title: 'Kung Fu',
	image: 'www.reddit.com',
	overview: 'Big Blue House',
	score: 7.8,
	id: 166428,
	isFavorite: false,
	releaseDate: "2015-01-03",},
	{	title: 'Spider Man',
	image: 'www.reddit.com',
	overview: 'Big Blue House',
	score: 9,
	id: 123455,
	isFavorite: false,
	releaseDate: "2017-01-03",},
]

export const mockReduxUser = {
	name: 'Jeff',
	id: 2,
}

export const mockReduxFavs = [
	{title: 'Spider Man',
	image: 'www.reddit.com',
	overview: 'Big Blue House',
	score: 9,
	id: 123455,
	isFavorite: true,
	releaseDate: "2017-01-03",}
]