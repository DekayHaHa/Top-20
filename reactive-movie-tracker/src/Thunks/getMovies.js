import { isLoading, hasError, addMovies } from '../actions'
import { cleanMovies } from '../utilities/cleaner'
import { APIkey } from "../utilities/key.js";

export const getMovies = () => {
	return async (dispatch) => {
		const url = `https://api.themoviedb.org/3/movie/popular?api_key=${APIkey}&language=en-US&page=1`;
		try {
			dispatch(isLoading(true))
			const response = await fetch(url)
			if (!response.ok) {
				throw Error(response.statusText)
			}
			dispatch(isLoading(false))
			const data = await response.json()
			const finalData = cleanMovies(data.results);
			dispatch(addMovies(finalData))
		} catch (error) {
			dispatch(hasError(error.message))
		}
	}
}