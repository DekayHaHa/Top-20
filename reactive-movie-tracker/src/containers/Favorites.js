import React, { Component } from "react";
import { cleanFavorites} from '../utilities/cleaner'
import Movie from "./Movie";
import { connect } from "react-redux";
import { retrieveAllFavorites} from '../utilities/api'
import { addMovies, addFavorites } from "../actions/index"


class Favorites extends Component {


}

export const mapStateToProps = state => ({
  movies: state.movies,
  activeUser: state.activeUser,
  favorites: state.favorites
});

export const mapDispatchToProps = dispatch => ({
  addMovies: (movies) => dispatch(addMovies(movies)),
  addFavorites: (movie) => dispatch(addFavorites(movie))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Favorites);
