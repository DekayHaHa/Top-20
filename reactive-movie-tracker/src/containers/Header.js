import React, { Component } from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { signOutUser } from "../actions/index";
import PropTypes from "prop-types";
import "../styles/Header.scss";
import { getMovies } from '../Thunks/getMovies'


export const Header = class extends Component {

  clearUserData = () => {
    this.props.signOutUser()
    this.props.getMovies()
  };

  render() {
    const { activeUser } = this.props
    let buttonText;
    let urlPath;
    let button;

    if (activeUser.id) {
      buttonText = 'Sign Out'
      urlPath = 'favorites'
      button = <button onClick={this.clearUserData} className="btn">{buttonText}</button>
    } else {
      buttonText = 'User Sign In'
      urlPath = 'login'
      button = < Link to = "/login" > <button className="btn">{buttonText}</button></Link >
    }

    const welcome = <p>Welcome Back, { activeUser.name }</p>
    return (
      <div className="header">
      <div>
        { activeUser.id > 0 && welcome }
        {button}
      </div>
        <Link to='/'><h2 className="title">Movie Tracker</h2></Link>
        <Link to={`/${urlPath}`}><button className="btn">Favorites</button></Link>
      </div>
    );
  }
};

Header.propTypes = {
  activeUser: PropTypes.object
};

export const mapStateToProps = state => ({
  activeUser: state.activeUser
});

export const mapDispatchtoProps = dispatch => ({
  signOutUser: () => dispatch(signOutUser()),
  getMovies: (url) => dispatch(getMovies(url))
});

export default connect(
  mapStateToProps,
  mapDispatchtoProps
)(Header);