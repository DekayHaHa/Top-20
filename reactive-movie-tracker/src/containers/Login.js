import React, { Component } from "react";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";
import PropTypes from "prop-types";
import "../styles/Login.scss";
import {createUser} from '../Thunks/createUser'
import { signIn } from '../Thunks/signIn'

export class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      password: "",
      email: ""
    };
  }

  newUser = e => {
    e.preventDefault();    
    const { createUser } = this.props
    const userInfo = {
      id: 1,
      name: this.state.name,
      password: this.state.password,
      email: this.state.email
    };
    createUser(userInfo)
  };

  handleSignIn = e => {
    e.preventDefault();
    const { signIn } = this.props
    const userInfo = {
      name: this.state.name,
      password: this.state.password,
      email: this.state.email
    };
    signIn(userInfo)
  };

  handleNameInput = event => {
    this.setState({
      name: event.target.value
    });
  };

  handlePasswordInput = event => {
    this.setState({
      password: event.target.value
    });
  };

  handleEmailInput = event => {
    this.setState({
      email: event.target.value
    });
  };

  signInInputs = () => {
    const { email, password } = this.state;
    return email && password ? false : true;
  };

  signUpInputs = () => {
    const { email, password, name } = this.state;
    return email && password && name ? false : true;
  };

  render() {
    const { email, password, name } = this.state;
    let formSubmitMethod = this.handleSignIn;
    const { error } = this.props
    if (error) {
      formSubmitMethod = this.handlePost;
    }
    const signInBtnToggle = this.signInInputs();
    const signUpBtnToggle = this.signUpInputs();
    return (
      <div className="Login">
        {this.props.activeUser.id > 0 && <Redirect to="/" />}
        {error && <h3 className="wrong-password">{error}</h3>}
        <form className="form" onSubmit={formSubmitMethod}>
          {error && (
            <label className="label">
              <input
                className='name-input'
                placeholder='Full Name'
                type="text"
                value={name}
                onChange={this.handleNameInput} />
            </label>
          )}
          <label label className="label">    
            <input
              className='email-input'
              placeholder='Email Address'
              type="text"
              value={email}
              onChange={this.handleEmailInput} />
          </label>
          <label label className="label">
            <input
              className='password-input'
              placeholder='Password'
              type="password"
              value={password}
              onChange={this.handlePasswordInput}
            />
          </label>
          <button className='btn signIn-btn' disabled={signInBtnToggle} onClick={this.handleSignIn}>
            Sign In
          </button>
          {error && (
            <button className='btn signUp-btn' disabled={signUpBtnToggle} onClick={this.newUser}>
              Sign Up NOW!
            </button>
          )}
        </form>
      </div>
    );
  }
}

Login.propTypes = {
  activeUser: PropTypes.object
};

export const mapStateToProps = state => ({
  activeUser: state.activeUser,
  error: state.error
});

export const mapDispatchtoProps = dispatch => ({
  createUser: (userInfo) => dispatch(createUser(userInfo)),
  signIn: (userInfo) => dispatch(signIn(userInfo)),
});

export default connect(
  mapStateToProps,
  mapDispatchtoProps
)(Login);
