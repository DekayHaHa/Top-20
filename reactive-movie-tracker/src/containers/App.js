import React, { Component } from "react";
import MovieHolder from "./MovieHolder";
import Favorites from './Favorites'
import { connect } from "react-redux";
import Header from "./Header";
import { Route, Switch } from "react-router-dom";
import Login from "./Login";
import { MovieDetails } from "../components/MovieDetails";
import { PageNotFound } from "../components/PageNotFound";
import PropTypes from "prop-types";
import "../styles/App.scss";
import { getMovies } from '../Thunks/getMovies'

export class App extends Component {
  componentDidMount() {
    this.getMovieData();
  }
  getMovieData = () => {
    this.props.getMovies()
  };
  findMovieToRender = id => {
    const MovieToRender = this.props.movies.find(movie => movie.id === id);
    const movie = <MovieDetails {...MovieToRender} />;
    return MovieToRender ? movie : false;
  };
  render() {
    return (
      <div className="App">
        <Route path="/" component={Header} />
        <Switch>
          <Route exact path="/" component={MovieHolder} />
          <Route exact path="/login" component={Login} />
          <Route exact path="/favorites" component={Favorites} />
          <Route
            exact
            path="/movie/:id"
            render={({ match }) => {
              const movie = this.findMovieToRender(parseInt(match.params.id));
              return movie || <PageNotFound />;
            }}
          />
          <Route component={PageNotFound} />
        </Switch>
      </div>
    );
  }
}

App.propTypes = {
  movies: PropTypes.array,
  activeUser: PropTypes.object
};

export const mapStateToProps = state => ({
  movies: state.movies,
  activeUser: state.activeUser
});

export const mapDispatchtoProps = dispatch => ({
  getMovies: (url) => dispatch(getMovies(url))
});

export default connect(
  mapStateToProps,
  mapDispatchtoProps
)(App);
