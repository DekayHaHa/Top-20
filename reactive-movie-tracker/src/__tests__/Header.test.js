import React from 'react';
import { shallow } from "enzyme";
import { Header, mapDispatchtoProps, mapStateToProps } from '../containers/Header'
import { mockReduxFavs, mockReduxMovies, mockReduxUser } from '../utilities/mockTestData'
import { signOutUser } from '../actions/index'
import { getMovies } from '../Thunks/getMovies'
jest.mock('../Thunks/getMovies')

describe("Header", () => {
	let wrapper;
	let mockFunc = jest.fn()
	beforeEach(() => {
		wrapper = shallow(<Header
			addFavorites={mockFunc}
			signInUser={mockFunc}
			movies={mockReduxMovies}
			activeUser={mockReduxUser}
			favorites={mockReduxFavs}
			/>);
	});


	it("should match the snapshot with all data passed in correctly", () => {
		expect(wrapper).toMatchSnapshot();
	});

	it("should map state to props", () => {
		//setup
		const mockStore = {
			activeUser: { name: 'steve', id: 2 },
			movies: ['should not show'],
			favorites: ['should not show']
		}
		const expected = {
			activeUser: { name: 'steve', id: 2 },
		}
		//execution
		const mappedProps = mapStateToProps(mockStore)
		//expectation
		expect(mappedProps).toEqual(expected)
	});
	it("should map signOutUser to props", () => {
		//setup
		const mockDispatch = jest.fn()
		const actionToDispatch = signOutUser()
		//execution
		const mappedProps = mapDispatchtoProps(mockDispatch)
		mappedProps.signOutUser()
		//expectation
		expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
	});
	
	it("should map getMovies to props", () => {
		//setup
		const mockUrl = 'www.reddit.com'
		const mockDispatch = jest.fn()
		const actionToDispatch = getMovies(mockUrl)
		//execution
		const mappedProps = mapDispatchtoProps(mockDispatch)
		mappedProps.getMovies(mockUrl)
		//expectation
		expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
	});
});