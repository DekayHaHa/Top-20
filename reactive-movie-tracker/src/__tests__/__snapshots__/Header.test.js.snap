// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match the snapshot with all data passed in correctly 1`] = `
<div
  className="header"
>
  <div>
    <p>
      Welcome Back, 
      Jeff
    </p>
    <button
      className="btn"
      onClick={[Function]}
    >
      Sign Out
    </button>
  </div>
  <Link
    to="/"
  >
    <h2
      className="title"
    >
      Movie Tracker
    </h2>
  </Link>
  <Link
    to="/favorites"
  >
    <button
      className="btn"
    >
      Favorites
    </button>
  </Link>
</div>
`;
